packages/phone-number-privacy/combiner/src/bls/bls-cryptography-client.ts

```diff
@@ -12,50 +12,95 @@ function flattenSigsArray(sigs: Uint8Array[]) {
+  private get totalSignatures(): number {
+    return this.unverifiedSignatures.length + this.verifiedSignatures.length
+  }
+  private get allSignatures(): Uint8Array {
+    const allSigs = this.verifiedSignatures.concat(this.unverifiedSignatures)
+    const sigBuffers = allSigs.map((response) => Buffer.from(response.signature, 'base64'))
+    return flattenSigsArray(sigBuffers)
+  }
 
-    try {
-      await threshold_bls.partialVerifyBlindSignature(
-        Buffer.from(polynomial, 'hex'),
-        Buffer.from(blindedMessage, 'base64'),
-        sigBuffer
-      )
-      this.verifiedSignatures.push(sigBuffer)
-    } catch (err) {
-      logger.error({ url: serviceResponse.url }, ErrorMessage.VERIFY_PARITAL_SIGNATURE_ERROR)
-      logger.error(err)
-    }
+  public addSignature(serviceResponse: ServicePartialSignature) {
+    this.unverifiedSignatures.push(serviceResponse)
   }
 
   /**
    * Returns true if the number of valid signatures is enough to perform a combination
    */
-  public hasSufficientVerifiedSignatures(): boolean {
+  public hasSufficientSignatures(): boolean {
     const threshold = config.thresholdSignature.threshold
-    return this.verifiedSignatures.length >= threshold
+    return this.totalSignatures >= threshold
   }
 
   /*
    * Computes the BLS signature for the blinded phone number.
+   * Throws an exception if not enough valid signatures
+   * and drops the invalid signature for future requests using this instance
    */
-  public async combinePartialBlindedSignatures(): Promise<string> {
+  public async combinePartialBlindedSignatures(
+    blindedMessage: string,
+    logger?: Logger
+  ): Promise<string> {
+    logger = logger ?? rootLogger
     const threshold = config.thresholdSignature.threshold
-    if (!this.hasSufficientVerifiedSignatures()) {
-      const err = new Error(
-        `${ErrorMessage.NOT_ENOUGH_PARTIAL_SIGNATURES} ${this.verifiedSignatures.length}/${threshold}`
+    if (!this.hasSufficientSignatures()) {
+      const error = new Error(
+        `${ErrorMessage.NOT_ENOUGH_PARTIAL_SIGNATURES} ${this.totalSignatures}/${threshold}`
+      )
+      logger.error(error)
+      throw error
+    }
+    // Optimistically attempt to combine unverified signatures
+    // If combination fails, iterate through each signature and remove invalid ones
+    // We do this since signature verification incurs higher latencies
+    try {
+      const result = threshold_bls.combine(threshold, this.allSignatures)
+      return Buffer.from(result).toString('base64')
+    } catch (error) {
+      logger.error(error)
+      // Verify each signature and remove invalid ones
+      // This logging will help us troubleshoot which signers are having issues
+      const verifySigReqs = this.unverifiedSignatures.map((unverifiedSignature) => {
+        return this.verifySignature(blindedMessage, unverifiedSignature, logger!)
+      })
+      await Promise.all(verifySigReqs)
```

Comment:
Should we call out here that the verification of the signatures isn't actually async? Otherwise we might be misleading that it's concurrent right now when it isn't